Add object filter ratio

mathmaniac43 opened the pull request:
Per issue #2948 (take two, this time targeting the correct branch `release-0.11.0`)

I was not able to figure out how to run frigate in a dev. container on a recorded video the way I wanted to for testing, but am open to doing so if I can get a bit of help/instruction.

I also did my best to maintain existing style/phrasing throughout, but feel free to make adjustments/requests if you think it should be improved.

Thanks!

mathmaniac43 committed:
Add object ratio config parameters

Issue: #2948

mathmaniac43 committed:
Add config test for object filter ratios

Issue: #2948

NickM-27 commented:
Since you added a field to be model you will also need to add a migration class to the migrations folder. Should be able to go off of what's there, let me know if you have any questions on that.

blakeblackshear commented on frigate/config.py:
```diff
@@ -210,6 +210,14 @@ class FilterConfig(FrigateBaseModel):
+        title="Minimum ratio of bounding box's width/height for object to be counted.",
+    )
+    max_ratio: float = Field(
+        default=24000000,  # @todo: Is there a way to make this `math.inf` but still load from YAML?
```
I think this is a fine default. Objects can't be less than 1px tall and nobody is going to have a camera that is 24000000  pixels wide.

blakeblackshear commented on frigate/video.py:
```diff
@@ -38,6 +38,9 @@
 def filtered(obj, objects_to_track, object_filters):
     object_name = obj[0]
+    object_score = obj[1]
+    object_bounds = obj[2]
```
let's rename this to `object_box` instead of `bounds` for consistency.

blakeblackshear commented on frigate/video.py:
```diff
@@ -615,8 +626,14 @@ def process_frames(
                 for group in detected_object_groups.values():
 
                     # apply non-maxima suppression to suppress weak, overlapping bounding boxes
+                    bounds = o[2]  # xmin, ymin, xmax, ymax
```
Similarly `box` instead of `bounds`.

blakeblackshear commented:
Where are you getting stuck running in the dev container? You should be able to run `python3 -m frigate` in the terminal window.

I use this config for local dev:
```yaml
cameras:
  back:
    ffmpeg:
      inputs:
        - path: /lab/frigate/airport.mp4
          input_args: -re -stream_loop -1 -fflags +genpts
```

mathmaniac43 commented:
@blakeblackshear Python isn't my primary programming language and I haven't developed any packages before, so was not familiar with the `-m` flag, I bet that will get me moving, thanks.

mathmaniac43 committed:
Address review comments

- Accept `ratio` default
- Rename `bounds` to `box` for consistency
- Add migration for new field

Issue: #2948

mathmaniac43 commented:
@blakeblackshear I got a `No module named 'frigate.version'` error, which looks like I can resolve using the `Makefile`. However, the dev. container does not have `make` in it, so I am going to have to do it outside the container. May be worth considering adding `make` to the container in the future.

NickM-27 commented:
@mathmaniac43 I believe you are supposed to make outside and then start the container. That's what I've always done

mathmaniac43 committed:
Fix logical errors

- field migrations require default values
- `clipped` referenced the wrong index for region, since it shifted
- missed an inclusion of `ratio` for detections in `process_frames`
- revert naming `o[2]` as `box` since it is out of scope!

This has now been test-run against a video, so I believe the kinks are
worked out.

Issue: #2948

blakeblackshear commented:
Building outside the container is listed in the docs here: https://docs.frigate.video/contributing#core

The command is now just `make` instead of `make amd64_frigate`. Feel free to fix that in the docs in this PR too.

NickM-27 commented on frigate/models.py:
```diff
@@ -19,6 +19,7 @@ class Event(Model):
     box = JSONField()
     area = IntegerField()
     retain_indefinitely = BooleanField(default=False)
+    ratio = FloatField()
```
Now that I am on my computer, just leaving an official comment about the migration being needed.

mathmaniac43 replied:
@NickM-27 I added the migration script [here](https://github.com/blakeblackshear/frigate/pull/2952/files/2ad0ed9b8d74475e05c800eecd82bf643db580b5#diff-2546ea6e97f5943ff2ab27f9cfd917da1af7eb19d6cdafd6ca96236bf82b2438), are you saying that any default values should match both there and here?

NickM-27 replied:
Sorry didn't see it, a couple notes:
1. yes the default values should be added here as well
2. There is already a 007 migration so this should be updated to 008

mathmaniac43 committed:
Update contributing notes for `make`

Issue: #2948

mathmaniac43 committed:
Fix migration

- Ensure that defaults match between Event and migration script
- Deconflict migration script number (from rebase)

Issue: #2948

mathmaniac43 commented:
@NickM-27 Makes sense, just fixed!

blakeblackshear commented on frigate/video.py:
```diff
@@ -38,6 +38,9 @@
 
 def filtered(obj, objects_to_track, object_filters):
```
You should include the ratio object filter here. This prevents them from being tracked.

mathmaniac43 committed:
Filter objects out of ratio bounds

Issue: #2948

NickM-27 commented on migrations/008_add_object_filter_ratio.py:
```diff
@@ -0,0 +1,38 @@
+"""Peewee migrations -- 007_add_object_filter_ratio.py.
```
Need to update this comment with the new number and filename

NickM-27 replied:
My PR which had a new migration in it just got merged so you'll need to update this file to `009` as well

mathmaniac43 committed:
Update migration file to 009

Issue: #2948

mathmaniac43 committed:
Merge branch 'release-0.11.0' into add-object-filter-ratio

NickM-27 commented:
Looks good to me👌

mathmaniac43 commented:
@blakeblackshear Pinging to make sure this has not fallen off your radar. Thanks.

blakeblackshear commented:
Just busy with other things at the moment
